### ISSUE TITLE
binary or pip install

### ISSUE THREAD

**Issue by casperdcl** [conjuring/conjuring#20]
a binary `conjuring` package:

- may be installable via
  + `pip`
    * would automatically install `docker-compose`
    * would require `pip` on the host
    * may not require `git` on the host
  + a single downloadable script/url
    * may not require `git` on the host
- can print `docker` install instructions/link
- make use of a local `custom.y(a)ml/json.config` file and `custom` config directory
- make use of a remote config (i.e. url of github repo)?
- Overall would make this project even more similar to [jupyter/repo2docker](https://github.com/jupyter/repo2docker) (albeit of course a faster multi-user version)

# example use

running `conjuring` would automatically:

## option 1: (current) manual git repo with optional submodules

```bash
git clone conjuring
cd conjuring

#### optional: convert to submodule
cd custom
git init
git add *
git commit -m "initial commit"
cd ..
mv custom orig_custom
git rm -rf custom
mv orig_custom custom
git submodule add ./custom custom
git commit -m "make custom/ a submodule"
#### end optional

# TODO: add users/packages to custom/ based on optional arguments to `conjuring init`?

docker-compose up --build -d
```

## option 2: hidden "installed" git repo with executable on PATH, searches CWD for `custom`
```bash
CWD="${PWD}"
# make sure conjuring is up-to-date in ~/.conjuring
[ -d ~/.conjuring ] || git clone conjuring ~/.conjuring
cd ~/.conjuring
git reset --hard
git pull
# ensure $CWD/custom exists
[ -d "$CWD"/custom ] || mv custom "$CWD"/
rm -rf custom
ln -s "$CWD"/custom
docker-compose up --build -d
cd "$CWD"
```

installation could also be automated in a single script:

### bash
- no additional dependencies on linux
- may require e.g. `git-bash` on windows (easy to install)
```bash
curl -o ~/.local/bin/conjuring https://conjuring.github.io/conjuring.sh
```

### pip
- requires python on the host, sort of defeating the point on conjuring
or via `pip`

```bash
pip install conjuring
```

## option 3: docker image to do things
- no additional dependencies
- properly cross-platform
- docker "conjuring-utils" image creates config (e.g. #37's python scripts can be run in the container)
- the generated config can then be used to create and run the actual "conjuring" image

## option X: my musings

I think this is overall the best option

The end user will:

- install `git`
  + `sudo apt-get install git curl` (linux)
  + `brew install git curl` (mac)
  + https://git-scm.com/download (win, includes `git-bash` and `curl` etc.)
- (win) open `git-bash` terminal
- `curl -o conjuring https://raw.githubusercontent.com/conjuring/conjuring/master/src/conjuring.sh`
  + or e.g. `curl -o ~/.local/bin/conjuring https://conjuring.github.io/conjuring.sh`
- `./conjuring`

The contents of `conjuring` are below.

- cross-platform
- would support (future/separate issue) config generation (e.g. #37 user generation) using more shell scripts

```bash
#!/usr/bin/env bash
set -e

CWD="${PWD}"
echo "Ensuring conjuring is up-to-date in ~/.conjuring"
[ -d ~/.conjuring ] || git clone https://github.com/conjuring/conjuring ~/.conjuring
pushd ~/.conjuring
git reset --hard
git pull
#files="$(ls -d */) $(ls *.yml Dockerfile)"
files=""  # blank for all
echo "Installing conjuring ($files) in $CWD"
echo -n "Overwrite all, overwrite older, keep existing, or quit [a/o/E/q]? "
read overwrite
case "$overwrite" in
  [aA])
    git archive --format=tar HEAD $files | (cd "$CWD" && tar -x -f - )
    ;;
  [oO])
    git archive --format=tar HEAD $files | (cd "$CWD" && tar -x -f - --keep-newer-files)
    ;;
  [eE]|"")
    git archive --format=tar HEAD $files | (cd "$CWD" && tar -x -f - --skip-old-files)
    ;;
  [qQ])
    exit 0
    ;;
  *)
    echo "Unknown option; aborting"
    exit 1
    ;;
esac
popd

docker-compose up --build -d
```

**Comment by casperdcl** [comment]
> [...] I've drafted up a note describing a workflow that relies mostly on what we have already, but requires couple of additions (cli interface and the custom.json option). The note is on the wiki, here:
> 
> 
> https://github.com/conjuring/conjuring/wiki/CLI-workflow

-- @darribas

**Comment by darribas** [comment]
As for the installation, I'd make `pip install`able only our code, and require folks to have `docker` and `docker-compose` working on their machines. Technically, what we'd install would *not* require `docker` to work (see the [wiki note](https://github.com/conjuring/conjuring/wiki/CLI-workflow)) as it'd be mostly creating project folders, etc.

As a "side-effect", it'd make everything else sooo much easier :-)

**Comment by jreades** [comment]
Copying over my notes from the Luxembourg conference -- I don't think what has been outlined so far _quite_ gets at what Dani and I were thinking. Here's the idea:

1. We think it makes sense to separate the bits that have to do with compiling JupyterHub and user accounts for use _in Docker_ from the bits that have to do with setting up _hardware_. The former is generic, the latter is going to always have a degree of OS specificity that we don’t want to get too caught up in: I’d like to have a good guide for setting up a NUC-type machine running Ubuntu 18 and place that in a kind of ‘gallery’ to which people can refer when setting up a standalone box, but which we don’t have to support. Other people can contribute other guides (e.g. Windows, OSX, etc.).
2. We want to create a kind of script allowing the user to run commands like `conjure project <name>…` and `conjure users…` in order to set up a new project and configure the user accounts before building the container. I kind of like the idea of a `conjure build` wrapper to docker just to make things simple (especially for doing things like tagging the container automatically) but I think Dani is less keen. So let's pretend that I'm agnostic on this (we could always just say "Hey, docker isn't installed so you might want to fix that before running me..." :-) This tries to strike a balance between coherence---you run `conjure <cmd> <params>` for everything---and flexibility---if you want to run `docker build...` you can. Anyway...
3. We’ve created Python scripts and committed these back into GitHub (I _think_) that can do some of this but: a) they need further thought; and b) we have no idea how to create this kind of executable (I’m envisioning something that would/could sit in /usr/local/bin/, or that could add itself to the $PATH).
4. With this binary/executable approach, we’d envision the user running `conjure new <project>` command within another project (e.g. a git repo such as github.com/kingsgeocomp/geocomputation) rather than (as is currently the case) the other way around. The point is to avoid asking them to manually copy all of the files into custom/shared and custom/home_default since that way lies forgetting to update your Notebooks when building a new image. This also makes it much more explicit that there will be likely be multiple conjuring projects on a single machine that you might want to deploy in turn without having to think through whether you need to remove a lot of images or containers before launching. So the project name would inform the docker container name (e.g. `conjuring:geocomp` if I run `conjure new geocomp` and tell it to pull in material from my `geocomputation` repo, but `conjuring:ectqg` if I run `conjuring new ectqg` and tell it to pull in material from `ectqg` repo). An ordinary way to do this would be to to create the new project _in_ the relevant local copy of the repo and then you could even add the config files to the repo itself so that you can quickly reproduce the entire workshop on another machine!
5. I actually think you’ve got 90% of this sorted, but this way we don’t need people to fork or branch from the `conjuring` repo itself, it looks more of a deployable tool that they use _in_ their own repos. If I’m remembering what Dani said correctly then some of this can be managed with a custom JSON config that points to other directories in various ways: e.g. allows access to teacher-specified directories (such as `data` or `lectures` outside of the Hub-user’s home and shared directories). Basically: avoid duplication, avoid copying of files, and (when building) try to ensure that the data and notebooks are copied to custom/shared properly.

**Comment by szschaler** [comment]
For point 3 can’t you add a shebang at the start of the Python script to point the shell to the python interpreter? Then you should be able to just set the executable flag and move the file into /usr/local/bin or similar and the system should be able to figure out how to run it. That’s the standard Linux approach to scripting languages.

Steffen

Dr. rer. nat. Steffen Zschaler AHEA

Senior Lecturer
King's College London
Department of Informatics

Visiting Scientist
The Francis Crick Institute

Email szschaler@acm.org
Phone +44 (020) 7848 1513
WWW   www.steffen-zschaler.de

**Comment by jreades** [comment]
Yes — that’s what I’ve done, but in testing, but I wasn’t sure how robust that was as a solution. Guess Windows always going to be a pain.

Jon

--
m: 07976987392
e: jon@reades.com
s: jreades

**Comment by szschaler** [comment]
Point taken. I’ve seen people write .bat files for Windows systems. Of course, there will always be the dependency on python, which on Windows is far from guaranteed to be installed.

Steffen

Dr. rer. nat. Steffen Zschaler AHEA

Senior Lecturer
King's College London
Department of Informatics

Visiting Scientist
The Francis Crick Institute

Email szschaler@acm.org
Phone +44 (020) 7848 1513
WWW   www.steffen-zschaler.de

**Comment by casperdcl** [comment]
yes I'm definitely going to convert to pure batch & shell scripts as it's silly to expect people to install python to install conjuring to install python. Just need to determine what exactly the scripts will do.

**Comment by casperdcl** [comment]
I've updated the issue description some more.

Note (point 5) that (apart from shared mounted folders at container runtime) docker *cannot* access folders outside the build directory. So avoiding copying isn't always going to be possible.

**Comment by jreades** [comment]
Of course -- non-duplication is an ambition, not a requirement. In terms of building a copy of the conjuring directory structure in the user's working directory (from `conjure new <project>`) do you think that `git` is fully required, or could each release target a 'digest' or a tarball/zipball hosted in the repo? That would leave us only requiring `wget` or `curl`... or have I misunderstood the purpose?

**Comment by casperdcl** [comment]
yes but

1. git makes it much easier to efficiently update each time the script is run (`git pull`)
2. the procedure for installing `curl` on any platform would make installing `git` as well trivial. On windows in fact, just inatalling `git-bash` will install `curl`

**Comment by jreades** [comment]
Cool, good answer.

--

**Comment by casperdcl** [comment]
ooh I just thought of another greay reason: to manage config, `conjuring config <options>...` could simply call `git config --file .conjuringrc <options>...`. Would mean we won't need to write config read/write code!
